# Lab | Inferential statistics

In [2]:
import pandas as pd
import numpy as np
import scipy.stats as st

## One tailed t-test

In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file files_for_lab/machine.txt. Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?

In [17]:
file_name = 'files_for_lab/machine_new.txt'

data = pd.read_csv(file_name, 
                   engine='python',
                   sep=None,
                   header=0,
                   names=['new_machine', 'old_machine'])

In [18]:
data

new_machine  old_machine
0         42.1         42.7
1         41.0         43.6
2         41.3         43.8
3         41.8         43.3
4         42.4         42.5
5         42.8         43.5
6         43.2         43.1
7         42.3         41.7
8         41.8         44.0
9         42.7         44.1

## One-sided independent samples test

H-null: mean time of new machine <= mean time of old machine
H-alt: mean time of new machine is > mean time of old machine

In [22]:
mean_new = data.new_machine.mean()
mean_old = data.old_machine.mean()
new_std, old_std = data.new_machine.std(ddof=1), data.old_machine.std(ddof=1)
sp = ( len(data) - 1 ) * ( new_std**2 ) +  ( len(data) - 1 ) * ( old_std**2 )
sp /= ( len(data) + len(data) - 2)
sp = np.sqrt(sp)
r = np.sqrt( (1/len(data)) + (1/len(data)) )
t = ( mean_new - mean_old )/ (sp * r)

In [23]:
st.t.cdf(t,df = len(data.new_machine)+len(data.old_machine)-2)

0.0016055712503872579

The p_value 0.0016 is less than our alpha (0.05). So we should reject H-null and accept the H-alt that the new machine is faster than the old machine.



## Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.



H-null: defense score = attack score

H-alt: defense score != attack score

We will need to use a two-sided test.

In [24]:
#reading in data

file_name = "files_for_lab/pokemon.csv"

poke_df = pd.read_csv(file_name)
poke_df

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

In [26]:
poke_df.drop('#', axis=1, inplace=True)

In [29]:
poke_sample = poke_df[['Attack', 'Defense']].sample(30, random_state=1)
poke_sample['difference'] = poke_sample['Attack']-poke_sample['Defense']
poke_sample.head()

Attack  Defense  difference
8       104       78          26
510      92       75          17
175      46       34          12
735      50       58          -8
242     105       75          30

In [30]:
sample_diff_mean, sample_diff_std = poke_sample['difference'].mean(), poke_sample['difference'].std(ddof=1)
sample_diff_mean, sample_diff_std

(2.8, 36.32079294288603)

In [31]:
#getting the t-statistic
t = sample_diff_mean / ( sample_diff_std / np.sqrt(poke_sample.shape[0]) )
print("The mean of our samples differences is: {:.2f}".format(sample_diff_mean))
print("The standard deviation of our samples differences is: {:.2f}".format(sample_diff_std))
print("Our t statistics is: {:.2f}".format(t))

The mean of our samples differences is: 2.80
The standard deviation of our samples differences is: 36.32
Our t statistics is: 0.42


In [32]:
# comparing to pvalue
p_value = st.t.cdf(t, df = poke_sample.shape[0]-1)
p_value

0.6620194128841399

In [33]:
#using another method
st.ttest_rel(poke_sample['Attack'],poke_sample['Defense'])

TtestResult(statistic=0.4222438544846934, pvalue=0.6759611742317202, df=29)

$$P = 0.66 > alpha/2 = 0.025$$

The pvalue is greater than alpha/2. We do not reject the H-null. 

## ANOVA

### Part 1
In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on: - Null hypothesis - Alternate hypothesis - Level of significance - Test statistic - P-value - F table

Context
Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data `anova_lab_data.xlsx` file in the `files_for_lab` folder

- State the null hypothesis
- State the alternate hypothesis
- What is the significance level
- What are the degrees of freedom of model, error terms, and total DoF


### Part 2

- In this section, use Python to conduct ANOVA.
- What conclusions can you draw from the experiment and why?



## Part 1 answers

H-null: The means of the etching rate for each level of power are the same

H-alt: The means of the etching rate for each level of power are not the same

Significance level = 0.05

Degrees of freedom
- The shape of the F distribution depends on two sets of degrees of freedom:  $d_{1}=K-1$ and $d_{2}=N-K$
- d1 is the total number of groups in the dataset - 1 (for this example the groups are the different power levels)
- d2 is the total length of the dataset - the number of groups in the dataset

In [3]:
chips = pd.read_excel('files_for_lab/anova_lab_data.xlsx')

In [4]:
chips.head()

Power   Etching Rate
0  160 W          5.43
1  180 W          6.24
2  200 W          8.79
3  160 W          5.71
4  180 W          6.71

In [8]:
chips.columns

Index(['Power ', 'Etching Rate'], dtype='object')

The shape of the F distribution depends on two sets of degrees of freedom:  $d_{1}=K-1$ and $d_{2}=N-K$

In [19]:
# calculating the dfs
d1 = chips['Power '].nunique() - 1
d2 = len(chips) - chips['Power '].nunique()

print(d1)
print(d2)

2
12


## Part 2 answers

In [10]:
# getting the mean for each type of power
grouped_df = chips.groupby('Power ')['Etching Rate'].agg(etching_mean='mean', samples='size').reset_index()
grouped_df

Power   etching_mean  samples
0  160 W         5.792        5
1  180 W         6.238        5
2  200 W         8.318        5

### SST
SST is a measure of how much the mean for each group deviates from the global mean.

We add the squares of the deviations multiplied by the number of members of the group, then divide by the number of members minus 1.

### SSE
SSE is measure of how much each value in a group deviates from the group mean.

In [12]:
chips['Power '].unique()

array(['160 W', '180 W', '200 W'], dtype=object)

In [13]:
len(chips[chips['Power '] == '160 W'])

5

In [15]:
S2t = 0
# for each type of power
for power in chips['Power '].unique():
    # get the number of values for this power type
    ng = len(chips[chips['Power '] == power])
    # find the mean of the etching rate for this power type and subtract it from the global mean, then square
    # sum these values
    S2t += ng * ((chips[chips['Power ']==power]['Etching Rate'].mean() - chips['Etching Rate'].mean()) **2)

# divide by the total number of types minus 1
S2t /= (chips['Power '].nunique()-1)
print("The value of S2t is {:.2f}".format(S2t))

The value of S2t is 9.09


In [17]:
S2E = 0
#for each type of power
for power in chips['Power '].unique():
    #for each etching rate value of this power:
    for rate in chips[chips['Power ']==power]['Etching Rate']:
        #get the sum of the difference between the etching rate and the group etching rate mean, squared
        S2E += (rate - chips[chips['Power ']==power]['Etching Rate'].mean())**2

# length of dataframe (total number of rows, minus the number of subgroups
S2E /= len(chips) - chips['Power '].nunique()

print("The value of S2E is {:.2f}".format(S2E))

The value of S2E is 0.25


In [18]:
# compute F

F = S2t / S2E
print("The value of F is {:.2f}".format(F))

The value of F is 36.88


The probability of getting an F value lower or equal to our F is obtained with the CDF.

In [20]:
st.f.cdf(F, dfn=d1, dfd=d2)

0.9999924934157276

Thus, the probability to get any value smaller or equal to F

$$P(x \le F=36.88)= 0.9999$$

The opposite is given by

In [21]:
1 - st.f.cdf(F, dfn=d1, dfd=d2)

7.5065842723986975e-06

Therefore, the probability to get a value bigger than F is:

$$P(x > F) = 1 - P(x \le F) = 0.0001 < 0.05$$

In [23]:
# using scipy stats to do the one way ANOVA
one_way = st.f_oneway(chips[chips['Power ']== '160 W']['Etching Rate'],
                      chips[chips['Power ']== '180 W']['Etching Rate'],
                      chips[chips['Power ']== '200 W']['Etching Rate'])

print(one_way)

F_onewayResult(statistic=36.87895470100505, pvalue=7.506584272358903e-06)
